Use `ImportIRInst` to point to C++ source locations on Clang errors and warnings

bricknerb opened the pull request:
Introduce `ImportIRId::Cpp` and refer to clang source location in its `ImportIRInst`.

Part of #5245.

bricknerb committed:
Introduce a Clang diagnostic instruction and use it to point to C++ source locations on Clang errors and warnings

bricknerb committed:
Add a comment on `TryConvertClangDiagnosticLoc()`

bricknerb committed:
Fix clang-tidy: Use forward declaration for `ClangSourceLocation` instead of including ti in `toolchain/sem_ir/ids.h`

jonmeow commented on testing/file_test/file_test_base.cpp:
```diff
@@ -271,8 +271,8 @@ auto FileTestBase::GetLineNumberReplacements(
   return {{.has_file = true,
-           .re = std::make_shared<RE2>(
-               llvm::formatv(R"(({0}):(\d+)?)", llvm::join(filenames, "|"))),
+           .re = std::make_shared<RE2>(llvm::formatv(
+               R"(:\ (?:\./)?({0}):(\d+)?)", llvm::join(filenames, "|"))),
```
Note, per discussion on #5246, I think this should be reverted.

bricknerb replied:
Thanks, reverted.
Note that this reverted regex now catches lines from C++ diagnostics like

// CHECK:STDERR: In file included from ./indirect_error.h:2:

See https://github.com/carbon-language/carbon-lang/pull/5262/commits/7921e909baf16ef20c583f56e34eacb02bf5cf7b

jonmeow replied:
Yes, and I've already commented on what the right direction is, most recently at https://github.com/carbon-language/carbon-lang/pull/5246#issuecomment-2784301206

Maybe how the current format differs from the format that we want would be a good topic for the next meeting?

bricknerb replied:
SGTM.

jonmeow commented on toolchain/sem_ir/clang_source_location.h:
```diff
@@ -0,0 +1,22 @@
+// it can be transformed to a path and line number.
+struct ClangSourceLocation {
+  clang::SourceLocation source_location;
+  const clang::DiagnosticsEngine* diag_engine;
```
Why does each location need its own diagnostic engine? That is, why isn't this per-file instead of per-location?

jonmeow replied:
I see also this is stored as part of ASTUnit, so maybe we don't need additional pointers to it?

bricknerb replied:
There were two reasons I opted for that:
1. It avoids the implicit assumption that the same `DiagnosticsEngine` is used in both places (though there's also the question of ownership).
2. We don't have the `ASTUnit` in the context when `HandleDiagnostic()` is called since it's being called while the AST is generated.
I've changed that and replaced `ClangSourceLocation` with `clang::SourceLocation`, by delaying emitting the diagnostics after `buildASTFromCodeWithArgs()` is finished.

jonmeow replied:
We care about keeping these structures small, and minimizing RAM consumption. As a consequence, I think we need to not store the same pointer per location if there's a way to avoid it.

bricknerb replied:
Thanks!
Makes sense.

jonmeow commented on toolchain/sem_ir/typed_insts.h:
```diff
@@ -680,6 +680,16 @@ struct ExportDecl {
 };
 
+// Allows pointing to a source location in C++ in diagnostics.
+struct ClangDiagnostic {
+  static constexpr auto Kind = InstKind::ClangDiagnostic.Define<Parse::NodeId>(
+      {.ir_name = "clang_diag",
+       .constant_kind = InstConstantKind::Never,
+       .is_lowered = false});
+
+  ClangSourceLocationId clang_loc_id;
+};
```
I think we probably don't want distinct instructions for these. Instead, we should probably be storing locations in their own array.

bricknerb replied:
Done.

jonmeow reviewed (commented):
Had you considered an approach reusing ImportIRInst? That is:

- For each SemIR, we should only have one C++ file directly in use. As a consequence, we can probably give it a dedicated IR slot (and ImportIRId, similar to ApiForImpl).
- For clang source locations, add a ValueStore similar to others, e.g. mapping a ClangSourceLocationId to a ClangSourceLocation (or just a `clang::SourceLocation`, per comment on that struct).
  - This part you seem to be close to with `clang_source_location_ids`, although the name should be `clang_source_locations`
- With that dedicated IR slot, reuse the ImportIRInst structure: but the InstId can in essence become a ClangSourceLocationId based on what the ImportIRId is.

I thought I'd brought up using ImportIRInst before, but looking at the `ClangDiagnostic` instruction, maybe it was unclear what I meant? Note a LocId is a root for SemIRLoc: we don't need an InstId, which I think would remove the need for ClangDiagnostic instructions.

jonmeow commented on toolchain/sem_ir/formatter.cpp:
```diff
@@ -1414,6 +1414,13 @@ class FormatterImpl {
+        sem_ir_->clang_source_location_ids().Get(id);
+    clang_source_location.source_location.print(
+        out_, clang_source_location.diag_engine->getSourceManager());
+  }
```
Sorry, I said I'd comment about formatting... Note that the IR is intended to have an IR-like syntax, so we typically will use things like quote-delimited strings (for the filename), and numbers. But also note we do things like `loc<line>_<col>` for line-of-code information: that's our way of transforming the information into a valid identifier.

bricknerb replied:
Used `loc<line>_<col>` format since I didn't know if and how you meant to add the filename there.

bricknerb committed:
Remove `diag_engine` from `ClangSourceLocation` and postpone emitting diagnostics until ASTUnit is set in `SemIR::File`

bricknerb committed:
Remove `ClangSourceLocation` and use `clang::SourceLocation` directly

bricknerb committed:
Format Clang source location as loc<line>_<col>

bricknerb committed:
Relax the regex for find anchoring source locations back (revert the change)

bricknerb committed:
Rename `clang_source_location_ids()` to `clang_source_locations()`

bricknerb committed:
Use `ImportIRInst` for Clang source locations instead of adding a dedicated instruction.
Add `ImportIRId::Cpp` similar to `ImportIRId::ApiForImpl`.
Change `ImportIRInst` to contain `ClangSourceLocationId` and not `InstId` for `ImportIRId::Cpp`.

bricknerb commented:
> Had you considered an approach reusing ImportIRInst? That is:
> 
> * For each SemIR, we should only have one C++ file directly in use. As a consequence, we can probably give it a dedicated IR slot (and ImportIRId, similar to ApiForImpl).
> * For clang source locations, add a ValueStore similar to others, e.g. mapping a ClangSourceLocationId to a ClangSourceLocation (or just a `clang::SourceLocation`, per comment on that struct).
>   
>   * This part you seem to be close to with `clang_source_location_ids`, although the name should be `clang_source_locations`
> * With that dedicated IR slot, reuse the ImportIRInst structure: but the InstId can in essence become a ClangSourceLocationId based on what the ImportIRId is.
> 
> I thought I'd brought up using ImportIRInst before, but looking at the `ClangDiagnostic` instruction, maybe it was unclear what I meant? Note a LocId is a root for SemIRLoc: we don't need an InstId, which I think would remove the need for ClangDiagnostic instructions.

Done.

jonmeow commented on toolchain/sem_ir/formatter.cpp:
```diff
@@ -1414,6 +1414,17 @@ class FormatterImpl {
+            clang_source_location);
+
+    out_ << "loc" << presumed_loc.getLine() << "_" << presumed_loc.getColumn();
+  }
```
Is this used? My suspicion would be that as a consequence of dropping the instruction, it might not be; but maybe I'm overlooking a path to reach it?

bricknerb replied:
Right, removed.

jonmeow commented on toolchain/sem_ir/file.h:
```diff
@@ -221,6 +222,13 @@ class File : public Printable<File> {
   }
   auto insts() -> InstStore& { return insts_; }
   auto insts() const -> const InstStore& { return insts_; }
+  auto clang_source_locations() -> ValueStore<ClangSourceLocationId>& {
+    return clang_source_locations_;
+  }
+  auto clang_source_locations() const
+      -> const ValueStore<ClangSourceLocationId>& {
+    return clang_source_locations_;
+  }
```
We've generally kept accessors in the same order as the fields, can you do the same here?

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_ref.h:
```diff
@@ -11,10 +11,15 @@
 
-// Sets the IR for ImportIRId::ApiForImpl. Should be called before AddImportIR
-// in order to ensure the correct ID is assigned.
+// Sets the IR for `ImportIRId::ApiForImpl`. Should be called before
+// `AddImportIR` and `SetCppImportIR` in order to ensure the correct ID is
+// assigned.
 auto SetApiImportIR(Context& context, SemIR::ImportIR import_ir) -> void;
 
+// Sets the IR for `ImportIRId::Cpp`. Should be called before `AddImportIR` and
+// after `SetApiImportIR` in order to ensure the correct ID is assigned.
+auto SetCppImportIR(Context& context, SemIR::ImportIR import_ir) -> void;
```
Had you considered having a single function that takes both IRs, to avoid adding ordering constraints?

In particular, I'm also wondering if the C++ import IR even needs to have anything passed in: per my other comment about leaving `sem_ir` as null on it, maybe it could always have the same structure?

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -121,13 +146,21 @@ class CarbonClangDiagnosticConsumer : public clang::DiagnosticConsumer {
   // The location that triggered calling Clang.
   SemIRLoc loc_;
+
+  struct ClangDiagnosticInfo {
+    clang::DiagnosticsEngine::Level level;
+    SemIR::ImportIRInstId import_ir_inst_id;
+    std::string message;
+  };
+
+  llvm::SmallVector<ClangDiagnosticInfo> diagnostics_infos_;
```
Should there be comments to explain this?

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/BUILD:
```diff
@@ -41,6 +41,7 @@ cc_library(
         "//toolchain/base:value_ids",
         "//toolchain/diagnostics:diagnostic_emitter",
         "//toolchain/parse:node_kind",
+        "@llvm-project//clang:basic",
```
I'm not immediately noticing any #include additions to match the new dep, where should I be looking?

bricknerb replied:
Removed, probably leftover from a previous iteration.
Thanks!

jonmeow commented on toolchain/sem_ir/inst_fingerprinter.cpp:
```diff
@@ -60,6 +60,10 @@ struct Worklist {
     AddString(sem_ir->identifiers().Get(ident_id));
   }
 
+  auto Add(ClangSourceLocationId /*clang_source_location_id*/) -> void {
+    CARBON_FATAL("Should never fingerprint a location");
+  }
+
```
What made this come up? Is removing it a build error?

bricknerb replied:
Seems like this was only an issue in a previous iteration.
Removed.

jonmeow commented on toolchain/sem_ir/id_kind.h:
```diff
@@ -152,6 +152,7 @@ using IdKind = TypeEnum<
     AssociatedConstantId,
     BoolValue,
     CallParamIndex,
+    ClangSourceLocationId,
```
Can this be removed? IdKind is for ids which are on instructions (see the type comment)

bricknerb replied:
Done.

jonmeow commented on toolchain/check/diagnostic_emitter.cpp:
```diff
@@ -12,6 +14,11 @@ namespace Carbon::Check {
 
 auto DiagnosticEmitter::ConvertLoc(SemIRLoc loc, ContextFnT context_fn) const
     -> Diagnostics::ConvertedLoc {
+  auto converted_clang_loc = TryConvertClangDiagnosticLoc(loc);
+  if (converted_clang_loc) {
+    return *converted_clang_loc;
+  }
```
Had you considered making `ConvertLocImpl` able to handle these locations, rather than something separate? Maybe specifically handling it as part of `GetAbsoluteNodeId`?

There are few things I'm thinking about here:

1. `ConvertLocImpl` is what's producing the "in import" output. I think we should have something similar for C++ imports. That's how we get the first line for things like:

    ```
    fail_import_cpp_file_with_multiple_errors.carbon:3:1: note: in import
    In file included from fail_import_cpp_file_with_multiple_errors.carbon.generated.cpp_imports.h:1:
    ./multiple_errors.h:3:2: error: "error2"
        3 | #error "error2"
          |  ^
    ```

2. This approach only handles direct uses of C++ locations in the same file. We might want to handle indirect locations (i.e., file A imports file B has a C++ reference), which would need the kind of indirect location handling that `GetAbsoluteNodeId` is providing.

3. We also use `GetAbsoluteNodeId` for debug info code locations, and we may want to produce debug info for imported C++ code locations.

bricknerb replied:
That was actually my first approach.
However, it seems like the entire logic of `ConvertLocImpl()` is about calculating a list of `AbsoluteNodeId` and processing that list.
Since I don't have `NodeId` for C++ code, it didn't seem to be a good fit.
WDYT?

jonmeow replied:
The reason I mention `GetAbsoluteNodeId` is because I think it would integrate with the three `GetAbsoluteNodeId` use-cases I mentioned. You're right that it's currently only a `NodeId`, but similar to `ImportIRInstId`, I think that should be a `ClangSourceLocID` based on the `IRId`

That said, I'd be okay with checking in as-is, with a TODO to switch approaches and eliminate `TryConvertClangDiagnosticLoc`.

bricknerb replied:
Added a TODO.

jonmeow commented on toolchain/sem_ir/import_ir.h:
```diff
@@ -29,19 +29,34 @@ struct ImportIR : public Printable<ImportIR> {
+
+    // Set iff `ir_id == ImportIRId::Cpp`.
+    ClangSourceLocationId clang_source_location_id;
+  };
```
Rather than a union, had you considered accessors that verify the right call is in use? e.g.:

```
  auto inst_id() -> InstId {
    CARBON_CHECK(ir_id != ImportIRId::Cpp);
    return InstId(raw_loc_index);
  }

  auto clang_source_location_id() -> ClangSourceLocationId {
    CARBON_CHECK(ir_id == ImportIRId::Cpp);
    return ClangSourceLocationId(raw_loc_index);
  }

  int32_t raw_loc_index;
```

That could also be matched with constructors doing equivalent verification.

bricknerb replied:
I've converted the struct to a class to make sure it's used correctly.
I've kept the `union`, but that's implementation detail now and it seems more type safe than `int32_t`.

jonmeow commented on toolchain/sem_ir/ids.h:
```diff
@@ -732,6 +744,14 @@ struct TypeId : public IdBase<TypeId> {
 };
 
+// The ID of a Clang Source Location.
+struct ClangSourceLocationId : public IdBase<ClangSourceLocationId> {
+  static constexpr llvm::StringLiteral Label = "clang_source_location";
```
```suggestion
struct ClangSourceLocId : public IdBase<ClangSourceLocId> {
  static constexpr llvm::StringLiteral Label = "clang_source_loc";
```

While "Location" matches clang, in our code we've been abbreviating it ([per idioms](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/idioms.md#abbreviations-used-in-the-code-aka-carbon-abbreviation-decoder-ring)). Maybe this should abbreviate too? (and actually, similar elsewhere you're spelling out "location" for Carbon)

bricknerb replied:
Done.

jonmeow commented on toolchain/check/check_unit.cpp:
```diff
@@ -120,6 +124,15 @@ auto CheckUnit::InitPackageScopeAndImports() -> void {
+  SetCppImportIR(context_,
+                 {.decl_id = SemIR::InstId::None,
+                  .is_export = false,
+                  .sem_ir = unit_and_imports_->cpp_import_names.empty()
+                                ? nullptr
+                                : &context_.sem_ir()});
```
Why does this assign a value to sem_ir instead of leaving it as null?

bricknerb replied:
This is used to verify that the `SemIRLoc` refers to `ImportIR` that was generated with the same `File` address.
If that's not the case, we could end up in a bad state.
I understand you might consider this check to be too costly, so I removed it.

jonmeow replied:
It's more, having a value seemed like it would create worse failures. With null, an attempt to read through it should be a seg fault. With a self-pointer, an attempt to read through it would work and might create mysterious errors or infinite loops. So my preference for null is that it seemed like it'd make bad code clearer, earlier.

Anyways, I think this is resolved, just detailing the difference in why I was commenting.

bricknerb replied:
Thanks!

jonmeow commented on toolchain/check/check_unit.cpp:
```diff
@@ -120,6 +124,15 @@ auto CheckUnit::InitPackageScopeAndImports() -> void {
                    {.decl_id = SemIR::InstId::None, .sem_ir = nullptr});
   }
 
+  // Add an implicit `Cpp` import If there is an implicit `api` import, set it
+  // second so that it uses the `ImportIRId::Cpp` when processed for imports.
```
Comment only partially replaced?

bricknerb replied:
Done.

jonmeow commented on toolchain/check/diagnostic_emitter.cpp:
```diff
@@ -56,6 +63,36 @@ auto DiagnosticEmitter::ConvertLocImpl(SemIRLoc loc,
+  clang::PresumedLoc presumed_loc =
+      sem_ir_->cpp_ast()->getSourceManager().getPresumedLoc(clang_loc);
+
+  return Diagnostics::ConvertedLoc{
```
We use last_byte_offset to order diagnostics. Can you please either add it based on the `import Cpp` location, or add a TODO to that effect?

You can see this for example by adding a test with a lexer error, such as `library "[[@TEST_NAME]]";  // Trailing comment` (before a `Cpp` import). This lexer error should allow parse/check to continue, and its use of `last_byte_offset` will cause it to come *after* this diagnostic until the `last_byte_offset` is set.

The desired result would be that the diagnostics that occur due to a `Cpp` import should have their `last_byte_offset` set to something sensible for the import that caused the `Cpp` handling (often, that's handled through `last_token` in `ConvertLoc`, and adjusting that could be one way of handling ordering).

bricknerb replied:
I've added a test and TODOs.

jonmeow reviewed (commented):
Thanks for the updates! For comments, if you'd like to add TODOs for bigger amounts of work [particularly around ConvertLoc comments], I think that'd be fine -- I think this is generally heading in a good direction.

bricknerb committed:
Use `loc` instead of `location`.
https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/idioms.md#abbreviations-used-in-the-code-aka-carbon-abbreviation-decoder-ring

bricknerb committed:
Remove `ClangSourceLocId` `IdKind`

bricknerb committed:
Remove Worklist::Add(ClangSourceLocId) overload

bricknerb committed:
Remove unnecessary dependency @llvm-project//clang:basic from typed_insts

bricknerb committed:
Remove unused `FormatName(ClangSourceLocId)`

bricknerb committed:
Move `clang_source_locs()` to match the field order

bricknerb committed:
Don't set `sem_ir` field in `ImportIR` for the `Cpp` implicit API import.

bricknerb committed:
Fix the comment on the `SetCppImportIR()` call

bricknerb committed:
Rename `diagnostics_infos_` to `diagnostic_infos_`

bricknerb committed:
Add comments on `ClangDiagnosticInfo` and `diagnostic_infos_`

bricknerb committed:
Add `fail_import_cpp_library_lexer_error.carbon` test that demonstrates the Lexer error wrongly comes after the Clang diagnostic.
Explicitly set `last_byte_offset` to zero and add TODOs to fix that.

bricknerb committed:
Convert `ImportIRInst` from a struct to a class and make sure that when creating and accessing `inst_id` is used when `ir_id` is not `ImportIRId::Cpp`. Otherwise `clang_source_loc_id` is used.

bricknerb committed:
Merge `SetCppImportIR()` with `SetApiImportIR()` and call it `SetApiImportIRs()`. No need to pass `ImportIR` for Cpp since it's constant.

jonmeow commented on toolchain/sem_ir/import_ir.h:
```diff
@@ -29,19 +29,57 @@ struct ImportIR : public Printable<ImportIR> {
+// LocId. For `Cpp` import, points to a Clang source location.
+class ImportIRInst : public Printable<ImportIRInst> {
+ public:
+  // Constructor for non `Cpp` import.
```
```suggestion
  // Constructor for a non-`Cpp` import.
```

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/import_ir.h:
```diff
@@ -29,19 +29,57 @@ struct ImportIR : public Printable<ImportIR> {
+    CARBON_CHECK(ir_id != ImportIRId::Cpp);
+  }
+
+  // Constructor for `Cpp` import.
```
```suggestion
  // Constructor for a `Cpp` import.
```

bricknerb replied:
Done.

jonmeow commented on toolchain/sem_ir/import_ir.h:
```diff
@@ -29,19 +29,57 @@ struct ImportIR : public Printable<ImportIR> {
 // A reference to an instruction in an imported IR. Used for diagnostics with
-// LocId.
-struct ImportIRInst : public Printable<ImportIRInst> {
+// LocId. For `Cpp` import, points to a Clang source location.
```
Small grammar nit.
```suggestion
// LocId. For a `Cpp` import, points to a Clang source location.
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_ref.cpp:
```diff
@@ -32,16 +32,26 @@ static auto InternalAddImportIR(Context& context, SemIR::ImportIR import_ir)
 }
 
-auto SetApiImportIR(Context& context, SemIR::ImportIR import_ir) -> void {
+// Sets the IR for a specific ImportIRId. Should be called in the right order in
+// order to ensure the correct ID is assigned.
```
Maybe clarify where to find the right order?
```suggestion
// Sets the IR for a specific ImportIRId. Ordering of calls is significant; see `SetApiImportIRs`.
```

Or, perhaps:

```suggestion
// Adds a special-cased IR and verifies it received the correct ID.
```

bricknerb replied:
Done (2nd suggestion).

jonmeow commented on toolchain/check/check_unit.cpp:
```diff
@@ -104,20 +108,20 @@ auto CheckUnit::InitPackageScopeAndImports() -> void {
-  // If there is an implicit `api` import, set it first so that it uses the
-  // ImportIRId::ApiForImpl when processed for imports.
+  // Call `SetApiImportIRs()` to set `ImportIRId::ApiForImpl` and
+  // `ImportIRId::Cpp` first as required.
```
```suggestion
  // `ImportIRId::Cpp` first, as required.
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_ref.cpp:
```diff
@@ -32,16 +32,26 @@ static auto InternalAddImportIR(Context& context, SemIR::ImportIR import_ir)
+// order to ensure the correct ID is assigned.
+static auto SetImportIR(Context& context, SemIR::ImportIR import_ir,
+                        SemIR::ImportIRId expected_import_ir_id) -> void {
   auto ir_id = SemIR::ImportIRId::None;
   if (import_ir.sem_ir != nullptr) {
     ir_id = AddImportIR(context, import_ir);
   } else {
     // We don't have a check_ir_id, so add without touching check_ir_map.
     ir_id = InternalAddImportIR(context, import_ir);
   }
-  CARBON_CHECK(ir_id == SemIR::ImportIRId::ApiForImpl,
-               "ApiForImpl must be the first IR");
+  CARBON_CHECK(ir_id == expected_import_ir_id,
+               "Actual ImportIRId ($0) != Expected ImportIRId ({1})", ir_id,
+               expected_import_ir_id);
+}
+
+auto SetApiImportIRs(Context& context, SemIR::ImportIR import_ir) -> void {
```
Previously `SetApiImportIR` only set the API file's `ImportIR`, thus the name.

Similarly, `AddImportIR` is significantly different from `SetImportIR`: the former is used with any IR, whereas the latter is only used with the special `ImportIRId` values.

Maybe renaming this to `SetSpecialImportIRs` instead of `SetApiImportIRs` would help reflect that it's now more than API's IR? Similarly, perhaps use the singular form `SetSpecialImportIR` instead of `SetImportIR`?

bricknerb replied:
Done.

bricknerb committed:
Rename `SetApiImportIRs()` to `SetSpecialImportIRs()` and `SetImportIR()` to `SetSpecialImportIR()`

bricknerb committed:
Fixes for comments grammar.

bricknerb committed:
Rewrite `SetSpecialImportIR()` comment

bricknerb committed:
Add a TODO to remove `TryConvertClangDiagnosticLoc()`

jonmeow commented on toolchain/check/check_unit.cpp:
```diff
@@ -104,20 +108,20 @@ auto CheckUnit::InitPackageScopeAndImports() -> void {
 
-  // If there is an implicit `api` import, set it first so that it uses the
-  // ImportIRId::ApiForImpl when processed for imports.
+  // Call `SetApiImportIRs()` to set `ImportIRId::ApiForImpl` and
```
```suggestion
  // Call `SetSpecialImportIRs()` to set `ImportIRId::ApiForImpl` and
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_ref.h:
```diff
@@ -13,7 +13,7 @@ namespace Carbon::Check {
 // Sets the IRs for `ImportIRId::ApiForImpl` and `ImportIRId::Cpp`. Should be
 // called before `AddImportIR` in order to ensure the correct IDs are assigned.
-auto SetApiImportIRs(Context& context, SemIR::ImportIR import_ir) -> void;
+auto SetSpecialImportIRs(Context& context, SemIR::ImportIR import_ir) -> void;
```
Sorry, was looking at this again and thinking it's not clear how `import_ir` is used from the signature. Maybe just detail in comments:
 
```suggestion
// Sets the IRs for `ImportIRId::ApiForImpl` from `import_ir`, and
// `ImportIRId::Cpp` from defaults. Should be called before `AddImportIR` in
// order to ensure the correct IDs are assigned.
auto SetSpecialImportIRs(Context& context, SemIR::ImportIR import_ir) -> void;
```

bricknerb replied:
Done.

jonmeow reviewed (approved):
With the TODOs, I think this is a reasonable step forward. Let me know if you want another look after merging conflicts.

bricknerb committed:
Fix reference to `SetApiImportIRs()` that was renamed to `SetSpecialImportIRs()`

bricknerb committed:
Clarify `SetSpecialImportIRs()` behavior.

bricknerb committed:
Merge branch 'main' into diag2

bricknerb commented:
> With the TODOs, I think this is a reasonable step forward. Let me know if you want another look after merging conflicts.

Thanks, I think the conflicting changes were relatively simple so I'm not sure another iteration is necessary.

bricknerb committed:
Add NOLINT for using `clang` namespace for forward declaration.
This is to suppress `readability-identifier-naming`.

bricknerb committed:
Use `NOLINTNEXTLINE(readability-identifier-naming)` instead of `NOLINT` to be more specific.
